modules/azure/recoveryservices.go

```diff
@@ -0,0 +1,151 @@
+	list, err := GetRecoveryServicesVaultBackupPolicyListE(vaultName, resourceGroupName, subscriptionID)
+	require.NoError(t, err)
+	return list
+}
+
+// GetRecoveryServicesVaultBackupProtectedVMList returns a list of protected VM's on the given vault/policy.
+// This function would fail the test if there is an error.
+func GetRecoveryServicesVaultBackupProtectedVMList(t *testing.T, policyName, vaultName, resourceGroupName, subscriptionID string) map[string]backup.AzureIaaSComputeVMProtectedItem {
+	list, err := GetRecoveryServicesVaultBackupProtectedVMListE(policyName, vaultName, resourceGroupName, subscriptionID)
+	require.NoError(t, err)
+	return list
+}
+
+// RecoveryServicesVaultExists indicates whether a recovery services vault exists; otherwise false or error.
+func RecoveryServicesVaultExistsE(vaultName, resourceGroupName, subscriptionID string) (bool, error) {
+	_, err := GetRecoveryServicesVaultE(vaultName, resourceGroupName, subscriptionID)
+	if err != nil {
+		if ResourceNotFoundErrorExists(err) {
+			return false, nil
+		}
+		return false, err
+	}
+	return true, nil
+}
+
+// GetRecoveryServicesVaultE returns a vault instance.
+func GetRecoveryServicesVaultE(vaultName, resourceGroupName, subscriptionID string) (*recoveryservices.Vault, error) {
+	subscriptionID, err := getTargetAzureSubscription(subscriptionID)
+	if err != nil {
+		return nil, err
+	}
+
+	resourceGroupName, err2 := getTargetAzureResourceGroupName((resourceGroupName))
+	if err2 != nil {
+		return nil, err2
+	}
+
+	client := recoveryservices.NewVaultsClient(subscriptionID)
+	// setup auth and create request params
+	authorizer, err := NewAuthorizer()
+	if err != nil {
+		return nil, err
+	}
+
+	client.Authorizer = *authorizer
+	vault, err := client.Get(context.Background(), resourceGroupName, vaultName)
+	if err != nil {
+		return nil, err
+	}
+	return &vault, nil
+}
+
+// GetRecoveryServicesVaultBackupPolicyListE returns a list of backup policies for the given vault.
+func GetRecoveryServicesVaultBackupPolicyListE(vaultName, resourceGroupName, subscriptionID string) (map[string]backup.ProtectionPolicyResource, error) {
+	subscriptionID, err := getTargetAzureSubscription(subscriptionID)
+	if err != nil {
+		return nil, err
+	}
+
+	resourceGroupName, err2 := getTargetAzureResourceGroupName(resourceGroupName)
+	if err2 != nil {
+		return nil, err2
+	}
+
+	client := backup.NewPoliciesClient(subscriptionID)
+	// setup authorizer
+	authorizer, err := NewAuthorizer()
+	if err != nil {
+		return nil, err
+	}
+
+	client.Authorizer = *authorizer
+	listIter, err := client.ListComplete(context.Background(), vaultName, resourceGroupName, "")
+	if err != nil {
+		return nil, err
+	}
+
+	policyMap := make(map[string]backup.ProtectionPolicyResource)
+	for listIter.NotDone() {
+		v := listIter.Value()
+		policyMap[*v.Name] = v
+		err := listIter.NextWithContext(context.Background())
+		if err != nil {
+			return map[string]backup.ProtectionPolicyResource{}, err
```

Comment:
NIT: This should return `nil, err` to match the other error cases. Alternatively, this can also return the in progress `policyMap` map.

Reply:
the NextWithContext call does not return the tuple of policyMap, error.  It only returns an error if there is one.

Reply:
Hmm I'm not sure I follow.

From the looks of it, the `return` here applies to `GetRecoveryServicesVaultBackupPolicyListE`, so can't you do `return nil, err` here instead of initializing the empty map, just like you are doing in the other error conditions like https://github.com/gruntwork-io/terratest/pull/640/files/255d7266e385f588e5df3bc59d95178903507e94#diff-2540952f83b86371751a0ee827adb7e0e06535af99ff0d83f431cc6e40bbca4cR85?

Or alternatively, can this be `return policyMap, err`? This matches some function signatures in go like reading a file, where it returns all the data that it has processed up to this point. In some use cases, this can help with error recovery and clean up code.

Reply:
ah got it, I thought you were referring to the iterator itself.  Fixing now.

Reply:
@rguthriemsft Just making sure I understand correctly: Are you planning on making this change? Doesn't look like it was pushed.

Reply:
sorry was on vacation.  making these changes later today.